### ISSUE TITLE
All pre-release changelog on release.

### ISSUE THREAD

**Issue by cvalentin-dkt** [python-semantic-release/python-semantic-release#817]
Hello,

I'm trying to include the commits from the different prereleases in the changelog of the release, but I can't find how to do it.

An idea ?

**Comment by codejedi365** [comment]
An interesting concept, this might be a place for a jinja macro to preprocess the release history object that is passed to the environment of the template renderer and then loop through the resulting object.

I've personally never written a jinja macro but that would be where I would start.

**Comment by codejedi365** [comment]
After re-reading this, I'm not sure what is your intended outcome for the changelog? 

1. Consolidate all prereleases commits to be under the full release version without the prerelease tags as sections?

Or

2. Consolidate all prereleases commits under the full release version while maintaining each prerelease section?

**Comment by cvalentin-dkt** [comment]
What would be interesting is that each "pre-release" has its own change log (as currently).
And when i create the "release", all the previous commits of the different RCs appear in the github changelog.
=> https://github.com/{org}/{repo}/releases/tag/vX.X.X

Actually, just last commit beetwen last RC and release is appear.

**Comment by codejedi365** [comment]
I think we are talking past each other given the terminology. I was specifically talking about the generation of `CHANGELOG.md` but it seems you were more interested in the GitHub Release Notes (a per version section of the changes). 

My understanding is that right now, the release notes are properly generated for each pre-release but you would like to have the newest full tag release to include all of the commits between it and the last full release as if the prereleases never existed. Is this correct?

> Actually, just last commit beetwen last RC and release is appear.

In other words: At the current moment, only commits between the last prerelease and the new release are  posted in the new versions release notes. This is my understanding, yes.

**Comment by ibiscp** [comment]
I was hoping to have the same behaviour here, if we generate pre-releases constantly, and after some time we generate a release, the release note is empty, it would make sense to have an option to consolidate all the pre-releases changes in the release.

**Comment by codejedi365** [comment]
@ibiscp, thank you for providing your input

**Comment by codejedi365** [comment]
Still in the backlog but a valuable improvement nonetheless

**Comment by codejedi365** [comment]
Still need to dig in to see if this still happens with the new changelog templates I plan to release shortly. Still in backlog.

**Comment by codejedi365** [comment]
Link #555

**Comment by codejedi365** [comment]
Still in backlog

**Comment by Hamzah-Haider** [comment]
@codejedi365 I have the same need for the use-case described here and in (#837). I briefly read through the issues and read you mentioning that it is actually possible with a macro. Is this still the case? Mind sharing what it is? If not, any idea when this will be a possibility or what my best course of action would be? 

I want to only have the changelog and release show changes between full releases rather than showing them under pre-releases and having the full version section empty.

**Comment by Hamzah-Haider** [comment]
Thank you @codejedi365 for your detailed and informative response. I really appreciate it. 

I will look into the jinja templates with what you have mentioned here and try to hack something out. Otherwise, I will try to see if the CHANGELOG and releases can be modified after the fact by some hack again. I don't really understand the effect of what you described in your last sentence but I will investigate. 

If I get something actually working with the jinja macro I can post it here or somewhere for people's reference as a current workaround.

Looking forward to when this is eventually supported. Awesome work!

**Comment by codejedi365** [comment]
> I don't really understand the effect of what you described in your last sentence but I will investigate. 

Sorry, clearly I left a lot to interpretation there and it wasn't very clear. What I meant is that essentially you update the changelog manually. It's easy to build or modify a changelog after your PSR release but the corrected changelog is not available on the tag that is defined or in the distribution artifact (which can be confusing to consumers). As I personally would try to get a changelog prior to release, I was explaining how to do that. It would probably take a bit of CI work (if you run PSR via CI) to prevent an automatic release and instead trigger PSR after you have prepared the changelog. To minimize the amount of work you need to do manually, I recommend first using the changelog update mode and then running the version command locally with all the persistent mechanisms off, which results in building the changelog (and stamping the version into your source code). At this point you just need to manually modify the changelog by combining the sections from the pre-releases together. Commit this change and then allow your for your release to execute normally with the changelog generation turned off (`--no-changelog` cli option or `changelog: false` for the GitHub Action).

Note that there is also the `changelog` subcommand but it will not create a section for the next release only current releases. Any new commits will be placed in an Unreleased section. This is still at a point where you can modify the changelog manually but it requires building the version section manually. 

Good luck!

**Comment by codejedi365** [comment]
Still in backlog

**Comment by codejedi365** [comment]
No movement yet, still in backlog

**Comment by Ferdinanddb** [comment]
I would also be interested by this feature, as it makes things clearer imho.
